pkg/agent/flowexporter/connections/l7_listener.go

```diff
@@ -0,0 +1,215 @@
+// Copyright 2023 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connections
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/netip"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/controller/l7visibility"
+	"antrea.io/antrea/pkg/agent/flowexporter"
+	k8sutil "antrea.io/antrea/pkg/util/k8s"
+	"antrea.io/antrea/pkg/util/podstore"
+)
+
+const socketPath = "/var/run/suricata/suricata_eve.socket"
+
+type podL7VisibilityAttrGetter interface {
+	IsL7VisibilityRequested(podNN string, ingress bool) bool
+}
+
+// Define struct to hold the L7 flow JSON values
+type Http struct {
+	Hostname      string `json:"hostname"`
+	URL           string `json:"url"`
+	UserAgent     string `json:"http_user_agent"`
+	ContentType   string `json:"http_content_type"`
+	Method        string `json:"http_method"`
+	Protocol      string `json:"protocol"`
+	Status        int32  `json:"status"`
+	ContentLength int32  `json:"length"`
+}
+
+// Event JSON format is defined by Suricata
+// See https://docs.suricata.io/en/latest/output/eve/eve-json-format.html?highlight=HTTP%20event#event-types
+type JsonToEvent struct {
+	Timestamp   string     `json:"timestamp"`
+	FlowID      int64      `json:"flow_id"`
+	InInterface string     `json:"in_iface"`
+	EventType   string     `json:"event_type"`
+	VLAN        []int32    `json:"vlan"`
+	SrcIP       netip.Addr `json:"src_ip"`
+	SrcPort     int32      `json:"src_port"`
+	DestIP      netip.Addr `json:"dest_ip"`
+	DestPort    int32      `json:"dest_port"`
+	Proto       string     `json:"proto"`
+	TxID        int32      `json:"tx_id"`
+	HTTP        *Http      `json:"http"`
+}
+
+type L7Listener struct {
+	l7Events                  map[flowexporter.ConnectionKey]map[string]map[int32]interface{}
+	l7mut                     sync.Mutex
+	suricataEventSocketPath   string
+	podL7VisibilityAttrGetter podL7VisibilityAttrGetter
+	podStore                  podstore.Interface
+}
+
+func NewL7Listener(
+	l7VisibilityController *l7visibility.L7VisibilityController,
```

Comment:
The parameter type should be the interface `podL7VisibilityAttrGetter`, for least priviliege principle and ease of test, which is also the point why we define the interface.
The interface should be public as you will need to initialize the struct in other package.